lib/improver/ensemble_copula_coupling/ensemble_copula_coupling_utilities.py

```diff
@@ -0,0 +1,144 @@
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""
+This module defines the utilities required for Ensemble Copula Coupling
+plugins.
+
+"""
+import copy
+import numpy as np
+import random
+
+import cf_units as unit
+import iris
+from iris.exceptions import CoordinateNotFoundError
+
+
+def create_percentiles(no_of_percentiles, sampling="quantile"):
+    """
+    Function to create percentiles.
+
+    Parameters
+    ----------
+    no_of_percentiles : Int
+        Number of percentiles.
+    sampling : String
+        Type of sampling of the distribution to produce a set of
+        percentiles e.g. quantile or random.
+        Accepted options for sampling are:
+        Quantile: A regular set of equally-spaced percentiles aimed
+                  at dividing a Cumulative Distribution Function into
+                  blocks of equal probability.
+        Random: A random set of ordered percentiles.
+
+    For further details, Flowerdew, J., 2014.
+    Calibrating ensemble reliability whilst preserving spatial structure.
+    Tellus, Series A: Dynamic Meteorology and Oceanography, 66(1), pp.1-20.
+    Schefzik, R., Thorarinsdottir, T.L. & Gneiting, T., 2013.
+    Uncertainty Quantification in Complex Simulation Models Using Ensemble
+    Copula Coupling.
+    Statistical Science, 28(4), pp.616-640.
+
+    Returns
+    -------
+    percentiles : List
+        Percentiles calculated using the sampling technique specified.
+
+    """
+    if sampling in ["quantile"]:
+        percentiles = np.linspace(
+            1/float(1+no_of_percentiles),
+            no_of_percentiles/float(1+no_of_percentiles),
+            no_of_percentiles).tolist()
+    elif sampling in ["random"]:
+        percentiles = []
+        for _ in range(no_of_percentiles):
+            percentiles.append(
+                random.uniform(
+                    1/float(1+no_of_percentiles),
+                    no_of_percentiles/float(1+no_of_percentiles)))
+        percentiles = sorted(percentiles)
+    else:
+        msg = "The {} sampling option is not yet implemented.".format(
+            sampling)
+        raise ValueError(msg)
+    return percentiles
+
+
+def create_cube_with_percentiles(percentiles, template_cube, cube_data):
+    """
+    Create a cube with a percentile coordinate based on a template cube.
+    The resulting cube will have an extra percentile coordinate compared with
+    the input cube. The shape of the cube_data should be the shape of the
+    desired output cube.
+
+    Parameters
+    ----------
+    percentiles : List
+        Ensemble percentiles.
```

Comment:
There should be the same no of percentiles as the first dimension on the cube_data